### ISSUE TITLE
Command Buffer

### ISSUE THREAD

**Issue by genaray** [genaray/Arch#17]
It is not possible to Query and add, destroy or add/remove an entity at the same time. Such a operation breaks the query aswell as the underlying world. 

To solve this issue we need to "record" or "buffer" such critical changes during the query to transfer those to the world after the query has finished. This would also allow us to leverage the parallel queries to the next level.

A first API draft would look like this : 
```csharp
var cb = new CommandBuffer(world);
var entityWhichIsBuffered = cb.Create();
cb.Destroy(someExistingEntity);
cb.Set<T>(someEntity, cmp);
cb.Add<T>(someEntity, cmp);
cb.Remove<T>(someEntity);
```

## Questions & Problems

Structs with object references or objects themself can not be written to unmanaged memory... therefore we would probably need to find an efficient solution to write and record such a broad variation of different items and types.

```csharp
cb.Set<int>(someEntity, 1);
cb.Set<MyStruct>(someEntity, new MyStruct());
cb.Set...
```

Parallel Queries and multiple threads may write to the command buffer at the same time, this needs to work somehow. We could lock the cb and its internals, however then the parallelqueries will become a bottleneck and pretty much obsolete. 

How could we do this ?

**Comment by stefanomerotta** [comment]
For a generic command queue that handle concurrent writers and/or Reader you could chetk the [Channel](https://learn.microsoft.com/en-us/dotnet/core/extensions/channels)  class. 

For a more specific command queue with a known set of command at compile time you can use a byte array, like is done in other ecs frameworks, but you will have to handle concurrent writers with a sync, Like SpinLock or ReaderWriterLockSlim

**Comment by genaray** [comment]
> For a generic command queue that handle concurrent writers and/or Reader you could chetk the [Channel](https://learn.microsoft.com/en-us/dotnet/core/extensions/channels) class.
> 
> For a more specific command queue with a known set of command at compile time you can use a byte array, like is done in other ecs frameworks, but you will have to handle concurrent writers with a sync, Like SpinLock or ReaderWriterLockSlim

Thanks ! Im gonna look at that :D 

I also had a different variant in mind, where each thread writes to its own thread local queue... and the main thread waits for it to finish, then we could loop over all thread lists and play back the recorded operations.

This would be more efficient since the threads wont block each other since each one has its own list.

**Comment by genaray** [comment]
The first draft of the command buffers are out... however this is more tricky than i thought. 

I think it makes sense to let buffers operate on managed structs, since this will not restrict the user and offers more flexibility. Unmanaged structs are easy to write to unmanaged memory and thus great to create a command buffer system, however since we also wanna include managed structs... this will not work.

## Attempt 1 

My current attempt is a bit complicated... first of all there like 3 different command buffers... one for creation of new entities, one for modifying existing ones and one for destroying entities.

I did this on purpose since each of that category requires different architecture to batch those "commands" efficiently. Actually there are no "commands", instead there only those categorys and internal datastructures to represent the operation.

```csharp
var world = World.Create();
var buffer = new CreationBuffer(world);

var entity = buffer.Create(_group);
buffer.Set(in entity, new Transform{ X = 10 });
buffer.Set(in entity, new Rotation{ W = 10 });

var secondEntity = buffer.Create(_group);
buffer.Set(in secondEntity, new Transform{ X = 10 });
buffer.Set(in secondEntity, new Rotation{ W = 10 });

buffer.Playback();
```

The `CreationBuffer` for example manages an internal lightweight and small archetype to "buffer" entities and to play them back to the real world later on. It basically simulates the real world, without being connected to it directly.  There still some differences tho, that `CreationBuffer` archetype does not use chunks for example.

The `ModificationBuffer` is a bit more harder since we somehow need to group set operations per entity. I solved this by implementing a internal `sparseset`, i picked this architecture since it allows us to simulate entities without requiring real archetypes and furthermore they are pretty fast to "fill". 

The `DestructionBuffer` is just a... blocking list basically.

## Further ideas

However i think it would probably make sense to specialise them even further. The current approach uses dictionarys again which is actually kinda slow... furthermore its a bit hacky since types are mostly not known. The user mostly knows what system should buffer which entities or what types of data... so generics could probably come in handy again : `CreationBuffer<...>` or `ModificationBuffer<....>`, this would speed them up even more. Also their playback to the real world since the types are known during compile time.

**Comment by genaray** [comment]
## Update

CommandBuffers are now working and also thread safe in #111830f
I also added a few more parallel methods since a few were missing.

Lets take a look at one example...

```csharp
var creationBuffer = new CreationBuffer(world, 1024);
world.ParallelQuery(in desc, (in Entity entity) => {

   var entity = creationBuffer.Create(group); 
   entity.Set<Explosion>(...);
   entity.Set<Position>(...);
});
creationBuffer.Playback();
```

Pretty easy and actually straight forward. 
There 4 different buffers, each one has a different structure and is specialised for its own case. They actually act more like a shadow world instead of a real "command buffer"... more like an additional data structure which than can be transfered into the world itself. 

All of them can be acessed multithreaded and should only allocate memory once and since they batch the "commands", they are very fast to playback. Actually as fast as real operations.

The thread safety could probably still be improved, im using locks mostly but there actually better and faster ways to do this, however for now its working fine.

Gonna merge this to the main in a few sec.

**Comment by genaray** [comment]
Case closed with #24 ... 

## Future topics
- Less locking within the buffers ? 
- Specialised buffers for entity creation : `CreationBuffer<T,TT,TTT...>` ?
